rider/src/main/kotlin/com/jetbrains/rider/plugins/unity/run/DefaultRunConfigurationGenerator.kt

```diff
@@ -3,51 +3,51 @@ package com.jetbrains.rider.plugins.unity.run
 import com.intellij.execution.RunManager
 import com.intellij.execution.configurations.ConfigurationTypeUtil
 import com.intellij.execution.configurations.UnknownConfigurationType
-import com.jetbrains.rider.UnityProjectDiscoverer
-import com.jetbrains.rider.UnityReferenceListener
+import com.intellij.openapi.project.Project
+import com.jetbrains.rdclient.util.idea.LifetimedProjectComponent
+import com.jetbrains.rider.plugins.unity.UnityHost
 import com.jetbrains.rider.plugins.unity.run.configurations.UnityAttachToEditorAndPlayFactory
 import com.jetbrains.rider.plugins.unity.run.configurations.UnityAttachToEditorFactory
 import com.jetbrains.rider.plugins.unity.run.configurations.UnityDebugConfigurationType
+import com.jetbrains.rider.util.reactive.whenTrue
 
-class DefaultRunConfigurationGenerator(unityProjectDiscoverer: UnityProjectDiscoverer, runManager: RunManager) {
+class DefaultRunConfigurationGenerator(project: Project, runManager: RunManager, unityHost: UnityHost) : LifetimedProjectComponent(project) {
 
     val ATTACH_CONFIGURATION_NAME = "Attach to Unity Editor"
     val ATTACH_AND_PLAY_CONFIGURATION_NAME = "Attach to Unity Editor & Play"
 
     init {
-        unityProjectDiscoverer.addUnityReferenceListener(object: UnityReferenceListener {
-            override fun hasUnityReference() {
-                // Clean up the renamed "attach and play" configuration from 2018.2 EAP1-3
-                // (Was changed from a separate configuration type to just another factory under "Attach to Unity")
-                val toRemove = runManager.allSettings.filter {
-                    it.type is UnknownConfigurationType && it.name == "Attach to Unity Editor & Play"
-                }
-                for (value in toRemove) {
-                    runManager.removeConfiguration(value)
-                }
+        unityHost.model.hasUnityReference.whenTrue(componentLifetime) {
```

Comment:
It might be a good idea to move `unityHost.model.hasUnityReference` inside `UnityProjectDiscoverer`. That would mean `UnityProjectDiscoverer` was the one place to go to see if we're in a Unity project - either by looking at the folder structure, or based on a property from the backend.

This property is effectively the "truth" about if we're a Unity project or not. So perhaps we should look at our existing `isUnityProject` methods and make them look at `hasUnityReference`, rather than file structure? So we could have a range of functions - `isUnityProjectFolder` for when it's not loaded properly, `isUnityProjectQuick` for when it's based on folder structure (for Unity Explorer) and `isUnityProject` that also checks the backend value?

Reply:
UnityProjectDiscoverer works for Re# and Rider, while unityHost.model.hasUnityReference is only for Rider.

Reply:
Not sure about the second part about isUnityProjectQuick

Reply:
Just giving the current `isUnityProject` a name that indicates that it's based on folder structure, rather than "the truth". It _might_ be a false positive. If you gave someone the API right now, they wouldn't know the difference between `isUnityProject` and `hasUnityReference`.